doc/GUIDE.md

```diff
@@ -0,0 +1,910 @@
+
+    public static int y(MemorySegment struct) { ... } // 1
+    public static void y(MemorySegment struct, int fieldValue) { ... } // 1
+
+    public static MemorySegment asSlice(MemorySegment array, long index) { ... } // 5
+
+    public static long sizeof() { ... } // 3
+
+    public static MemorySegment allocate(SegmentAllocator allocator) { ... } // 4
+    public static MemorySegment allocateArray(long elementCount,
+            SegmentAllocator allocator) { ... } // 4
+
+    public static MemorySegment reinterpret(MemorySegment addr, Arena arena,
+            Consumer<MemorySegment> cleanup) { ... } // 6
+    public static MemorySegment reinterpret(MemorySegment addr, long elementCount,
+            Arena arena, Consumer<MemorySegment> cleanup) { ... } // 6
+}
+```
+
+There are:
+
+1. a getter and setter for each field of the struct, which takes a pointer to a struct
+  (a `MemorySegment`) to get/set the field from/to.
+2. meta-data accessors for each field (`xxx$layout()` and `xxx$offset()`).
+3. meta-data accessors `sizeof` and `layout`, which can be used to get the size and layout
+  of the struct.
+4. `allocate*` methods for allocating a single struct or arrays of structs.
+5. an `asSlice` method which can be used to access elements of an array of structs.
+6. two `reinterpret` methods which can be used to sanitize raw addresses
+  returned by native code, or read from native memory.
+
+The following example shows how to allocate a struct using the `allocate` method, and then
+sets both the `x` and `y` field to `10` and `5` respectively:
+
+```java
+// Main.java
+
+try (Arena arena = Arena.ofConfined()) {
+    MemorySegment point = Point.allocate(arena);
+    Point.x(point, 10);
+    Point.y(point, 5);
+    // ...
+}
+```
+
+For working with arrays of structs, we can use the `allocateArray` method which accepts an
+additional element count, indicating the length of the array:
+
+```java
+// Main.java
+
+try (Arena arena = Arena.ofConfined()) {
+    int arrLen = 5;
+    MemorySegment points = Point.allocateArray(arrLen, arena);
+
+    for (int i = 0; i < arrLen; i++) {
+        MemorySegment element = Point.asSlice(points, i);
+        Point.x(element, 10 + i);
+        Point.y(element, 5 + i);
+    }
+
+    // ...
+}
+```
+
+In the above example, the `asSlice` method is used to _slice_ out a section of
+the array, which corresponds to a single `Point` struct element. This method
+can be used to access individual elements of the `points` array when given
+an index.
+
+Finally, the `reinterpret` method can be used to _sanitize_ a pointer that is
+returned from native code. Let's say we have a C function that creates an instance
+of a `Point`, and returns a pointer to it, as well as a function that deletes a
+point, given a pointer:
+
+```c
+struct Point* new_point(void);
+void delete_point(struct Point* ptr);
+```
+
+The pointer that is returned by the corresponding method that jextract generates
```

Comment:
Consider streamlining this by saying "The pointer that is returned by new_point does not have..."